### ISSUE TITLE
difference in result between machines (tried nthread)

### ISSUE THREAD

**Issue by germayneng** [Microsoft/LightGBM#1882]
Good day,

I realized that in different machines, I am not able to reproduce the exact same results from lgbm. In the same machine, running the sript multiple times give me the same results. So i do not think it is from the script itself. 

By setting up a EC2 in aws and creating a virtual venv and  installing the requirement,txt (so the modules are the same with my local virtual environment. ) 

There are some sort of randomness still in the prediction. Of course, using the aws and running multiple times gives the same results, except that it is simply different from the local one. 

In sum: i created 3 environment: local base, local venv and aws venv. Running the scripts give same results in the same environment but differ across environment. Seed are already set and nthread =1

Some params: 

> params = {
>     'nthread' : 1,
>     'learning_rate': 0.01,
>     'num_leaves':80,
>     'max_depth': 7,
>     'boosting': 'gbdt',
>     'objective': 'regression',
>     'metric': 'rmse',
>     'colsample_bytree': 0.5,
>     #'feature_fraction_seed': 1,
>     'subsample': 0.7,
>     'subsample_freq': 3,
>     'min_child_weight': 0,
>     'verbose': 0,
>     'min_gain_to_split': 3, # gamma
>     'boost_from_average':False,
>     'seed': 12}

and my requirement.txt to ensure lgbmm is the same version: 

> pandas==0.23.0
> numpy==1.14.3
> lightgbm==2.1.2
> pmdarima==1.0.0
> patsy==0.5.0
> statsmodels==0.9.0
> scipy==1.1.0
> tqdm==4.23.4
> scikit-learn==0.19.1

will be good to hear what may caused this. thanks!

**Comment by Laurae2** [comment]
Do you have exactly the same environment? Such as identical gcc, identical OS version, identical kernel.

**Comment by germayneng** [comment]
hi laurae,

for the case of aws vs local, i may not guarantee that i have the same os ... etcc but if it is base vs venv with the same lgbm version, will it guarantee the same results?

**Comment by Laurae2** [comment]
No, you need at least an identical OS and an identical compiler (and probably an identical kernel now due to cache trashing from security mitigations) to guarantee identical random number generation.

**Comment by germayneng** [comment]
Thank you for the prompt reply. 

So basically the random number generation is the factor that is differing across the environment? Anyway to ensure compiler and kernel to be the same?

**Comment by guolinke** [comment]
You can try the latest version as well. I remember we fix a bug for consistency.( std::sort -> std::stable_sort)
BTW, what is the result when removed sampling (row and col)?

**Comment by germayneng** [comment]
@guolinke  i believe 2.1.2 is the latest? Alsso, do you mean removing 

'colsample_bytree': 0.5,
'subsample': 0.7,
'subsample_freq': 3,

**Comment by guolinke** [comment]
The latest is 2.2.2

**Comment by germayneng** [comment]
Thanks @guolinke  and @Laurae2 

Managed to isolate the issue. It is not with lgbm. With the params as above, the result can be reproduced, even with lgbm 2.2.2 vs 2.1.2. 

Issue was another module: pmdarima which caused a slightly variation in float numbers but was hidden in between the dataset.
